### PR TITLE
Blood refactor stuff

### LINKED ISSUE #8094: Admin healing a slime person removes all slime jelly from them and drains hunger bar.
Rejuvenate removes all reagents, including slimepersons slime jelly. Due to the fact that slime people use their hunger bar to replenish jelly, slime person ends up with a red hunger bar and 29.5u of slime jelly instead of 100u after healing them.

### LINKED ISSUE #11895: Shoes don't lose the blood overlay once blood runs out
I'm pretty sure they're supposed to?

### LINKED ISSUE #13005: Slimepeople and golems duplicate blood when drinking/injecting it.
If a slimeperson or golems attempts to inject themselves with blood or even drink it, the blood will be both transfered into their body and remain in the container it was consumed from, effectively duplicating the blood. May or may not happen with other mutant races.

### LINKED ISSUE #17812: O- blood fucking poisions people
NOT EVEN O- BLOOD IS SAFE FROM BUGGY BLOOD TYPES

### PR DESCRIPTION
**Pull request by phil235**
- I made human/handle_blood() less shitty.
- We no longer leave a blood trail if blood_volume is below a certain level which depends on the brute damage received. This way dragging a wounded player doesn't have too much of a bad effect.
- We no longer give humans a second reagents var containing blood, and we don't put exotic blood in the mob's reagents. Now we don't store any blood substance inside the mob, we only have a blood_volume var. When we draw blood with syringe we create the reagent that match the type (blood reagent, or an exotic blood reagent)
- can't draw blood from mob if it's low on blood, so we can't empty a mob of its blood entirely.
- Removed the blood type preference appearing in character setting.
- blood pack automatically show blood type, unless manually labeled.
- removed bloody_hands_mob human var and same name glove var.
- Some animals now have blood (pets, goat, cows)
- hitting and dragging mobs with actual blood in their veins leaves blood on you and a trail on the floor.

Fixes  #17812
Fixes #13005
Fixes #11895
Fixes #8094

:cl: phil235
rscadd: Monkeys and all other animals that should have blood now has it. Beating them up will make you and your weapon bloody, just like beating a human does. Dragging them when wounded and lying will leave a blood trail. Their blood is still mostly cosmetic, they suffer no effects from low blood level, unlike humans.
rscadd: When a mob leaves a blood trail while dragged, it loses blood. You can no longer drag a corpse to make an inifinite amount of blood trails, because once the victim's blood reaches a certain threshold it no longer leaves a blood trail (and no longer lose any more blood). The threshold depends on how much damage the mob has taken. You can always avoid hurting the dragged mob by making them stand up or by buckling them to something or by putting them in a container.
rscdel: You can no longer empty a mob of its blood entirely with a syringe, once the mob's blood volume reaches a critically low level you are unable to draw any more blood from it.
tweak: A changeling absorbing a human now sucks all their blood.
/:cl:

### CODE REVIEW

File: code/game/machinery/computer/medical.dm [hunk 1]
```diff
@@ -563,35 +563,31 @@
...
+					if(5)
+						R.fields["p_stat"] = pick("*Unconcious*", "Active", "Physically Unfit")
+					if(6)
+						R.fields["m_stat"] = pick("*Insane*", "*Unstable*", "*Watch*", "Stable")
+				continue
+
+			else if(prob(1))
+				qdel(R)
+				continue
+	..()
```
Comment (phil235) [inline code/game/machinery/computer/medical.dm]: completely unrelated to the blood refactor, just simplifying the code.

File: code/game/machinery/iv_drip.dm [hunk 2]
```diff
@@ -4,7 +4,7 @@
 	icon_state = "iv_drip"
 	anchored = 0
 	density = 1
-	var/mob/living/carbon/human/attached = null
+	var/mob/living/carbon/attached = null
```
Comment (phil235) [inline code/game/machinery/iv_drip.dm]: you can harvest blood from monkeys with the iv drip.

File: code/game/objects/effects/decals/Cleanable/humans.dm [hunk 3]
```diff
@@ -3,7 +3,7 @@
 	desc = "It's red and gooey. Perhaps it's the chef's cooking?"
 	gender = PLURAL
 	density = 0
-	layer = ABOVE_OPEN_TURF_LAYER
+	layer = ABOVE_NORMAL_TURF_LAYER
```
Comment (phil235) [inline code/game/objects/effects/decals/Cleanable/humans.dm]: so it's not hidden when the blood decal is on a wall.

File: code/game/objects/effects/decals/cleanable.dm [hunk 4]
```diff
@@ -70,6 +70,8 @@
 				add_blood = bloodiness
 			bloodiness -= add_blood
 			S.bloody_shoes[blood_state] = min(MAX_SHOE_BLOODINESS,S.bloody_shoes[blood_state]+add_blood)
+			if(blood_DNA && blood_DNA.len)
+				S.add_blood(blood_DNA)
```
Comment (phil235) [inline code/game/objects/effects/decals/cleanable.dm]: you walk on a blood splatter, the dna info of the blood is transfered to your shoes.

File: code/game/objects/items/devices/scanners.dm [hunk 5]
```diff
@@ -183,26 +183,29 @@ MASS SPECTROMETER
...
+				if(H.bleed_rate)
+					user << "<span class='danger'>Subject is bleeding!</span>"
+			var/blood_percent =  round((C.blood_volume / BLOOD_VOLUME_NORMAL)*100)
+			var/blood_type = C.dna.blood_type
+			if(blood_id != "blood")//special blood substance
+				blood_type = blood_id
+			if(C.blood_volume <= BLOOD_VOLUME_SAFE && C.blood_volume > BLOOD_VOLUME_OKAY)
+				user << "<span class='danger'>LOW blood level [blood_percent] %, [C.blood_volume] cl,</span> <span class='info'>type: [blood_type]</span>"
+			else if(C.blood_volume <= BLOOD_VOLUME_OKAY)
+				user << "<span class='danger'>CRITICAL blood level [blood_percent] %, [C.blood_volume] cl,</span> <span class='info'>type: [blood_type]</span>"
```
Comment (phil235) [inline code/game/objects/items/devices/scanners.dm]: we detect blood levels of monkeys and humans with exotic blood.

File: code/game/objects/items/stacks/medical.dm [hunk 6]
```diff
@@ -40,7 +40,7 @@
 			if(H.bleedsuppress)
 				user << "<span class='warning'>[H]'s bleeding is already bandaged!</span>"
 				return
-			else if(!H.blood_max)
+			else if(!H.bleed_rate)
```
Comment (phil235) [inline code/game/objects/items/stacks/medical.dm]: blood_max...great var name, Pap.

File: code/modules/client/preferences.dm [hunk 7]
```diff
@@ -47,7 +47,6 @@ var/list/preferences_datums = list()
 	var/be_random_body = 0				//whether we'll have a random body every round
 	var/gender = MALE					//gender of character (well duh)
 	var/age = 30						//age of character
-	var/blood_type = "A+"				//blood type (not-chooseable)
```
Comment (phil235) [inline code/modules/client/preferences.dm]: blood type no longer appears in the player window during lobby, we couldn't even choose the blood type before and having this as preference would've made my changes harder to implement. so bye bye

File: code/modules/detectivework/footprints_and_rag.dm [hunk 8]
```diff
@@ -1,11 +1,9 @@
 
 /mob
 	var/bloody_hands = 0
-	var/mob/living/carbon/human/bloody_hands_mob
 
 /obj/item/clothing/gloves
 	var/transfer_blood = 0
-	var/mob/living/carbon/human/bloody_hands_mob
```
Comment (phil235) [inline code/modules/detectivework/footprints_and_rag.dm]: fuck those two vars in particular.

File: code/modules/mob/living/carbon/carbon.dm [hunk 9]
```diff
@@ -296,16 +299,6 @@
...
 
-
-/mob/living/carbon/getTrail()
-	if(getBruteLoss() < 300)
-		if(prob(50))
-			return "ltrails_1"
-		return "ltrails_2"
-	else if(prob(50))
-		return "trails_1"
-	return "trails_2"
```
Comment (phil235) [inline code/modules/mob/living/carbon/carbon.dm]: same code as living/getTrail() now

File: code/modules/mob/living/carbon/human/death.dm [hunk 10]
```diff
@@ -57,4 +57,5 @@
 /mob/living/carbon/proc/Drain()
 	ChangeToHusk()
 	disabilities |= NOCLONE
+	blood_volume = 0
```
Comment (phil235) [inline code/modules/mob/living/carbon/human/death.dm]: the changeling sucks out all your fluids!

File: code/modules/mob/living/carbon/human/human_movement.dm [hunk 11]
```diff
@@ -48,6 +48,8 @@
 						FP.blood_state = S.blood_state
 						FP.entered_dirs |= dir
 						FP.bloodiness = S.bloody_shoes[S.blood_state]
+						if(S.blood_DNA && S.blood_DNA.len)
+							FP.transfer_blood_dna(S.blood_DNA)
```
Comment (phil235) [inline code/modules/mob/living/carbon/human/human_movement.dm]: bloody footprints contain the same dna info as the splatter they come from.

File: code/modules/mob/living/carbon/human/species_types.dm [hunk 12]
```diff
@@ -246,6 +247,7 @@
 	if(callback)
 		callback.Remove(C)
 	C.faction -= "slime"
+	C.blood_volume = min(C.blood_volume, BLOOD_VOLUME_NORMAL)
```
Comment (phil235) [inline code/modules/mob/living/carbon/human/species_types.dm]: slime people blood volume can go super high (to split into two), so we reduce it to normal level if necessary on species change.

File: code/modules/mob/living/living.dm [hunk 13]
```diff
@@ -572,8 +573,8 @@ Sorry Giacom. Please don't be mad :(
 		if(get_dist(src, pulling) > 1 || ((pull_dir - 1) & pull_dir)) //puller and pullee more than one tile away or in diagonal position
 			if(isliving(pulling))
 				var/mob/living/M = pulling
-				if(M.lying && !M.buckled && (prob(M.getBruteLoss() / 2)))
-					makeTrail(T, M)
+				if(M.lying && !M.buckled && (prob(M.getBruteLoss()*200/M.maxHealth)))
```
Comment (phil235) [inline code/modules/mob/living/living.dm]: we need to use the ratio of brute damage to maxhealth because of living mobs with very low health.

File: code/modules/mob/living/simple_animal/simple_animal.dm [hunk 14]
```diff
@@ -411,6 +411,7 @@
 		icon_state = icon_dead
 		stat = DEAD
 		density = 0
+		lying = 1
```
Comment (phil235) [inline code/modules/mob/living/simple_animal/simple_animal.dm]: I hope this doesn't cause issues, this is needed so animals with blood leave a blood trail when dead.

File: code/modules/reagents/chemistry/holder.dm [hunk 15]
```diff
@@ -142,10 +142,6 @@ var/const/INJECT = 5 //injection
 	var/trans_data = null
 	for(var/reagent in reagent_list)
 		var/datum/reagent/T = reagent
-		if(T.id == "blood" && ishuman(target))
-			var/mob/living/carbon/human/H = target
-			H.inject_blood(my_atom, amount)
-			continue
```
Comment (phil235) [inline code/modules/reagents/chemistry/holder.dm]: Fuck those four lines in particular.

File: code/modules/reagents/chemistry/reagents/other_reagents.dm [hunk 16]
```diff
@@ -1,22 +1,32 @@
...
 /datum/reagent/blood
-			data = list("donor"=null,"viruses"=null,"blood_DNA"=null,"blood_type"=null,"resistances"=null,"trace_chem"=null,"mind"=null,"ckey"=null,"gender"=null,"real_name"=null,"cloneable"=null,"factions"=null)
-			name = "Blood"
-			id = "blood"
-			color = "#C80000" // rgb: 200, 0, 0
+	data = list("donor"=null,"viruses"=null,"blood_DNA"=null,"blood_type"=null,"resistances"=null,"trace_chem"=null,"mind"=null,"ckey"=null,"gender"=null,"real_name"=null,"cloneable"=null,"factions"=null)
+	name = "Blood"
+	id = "blood"
+	color = "#C80000" // rgb: 200, 0, 0
+	metabolization_rate = 5 //fast rate so it disappears fast.
```
Comment (phil235) [inline code/modules/reagents/chemistry/reagents/other_reagents.dm]: fast metabolization is required, because injecting blood into a mob to replenish his blood volume doesn't automatically delete the reagent, so it's probably better to get rid of the blood quickly (and nothing is lost since blood has no on_mob_life().

File: code/modules/reagents/chemistry/reagents/other_reagents.dm [hunk 17]
```diff
@@ -1,22 +1,32 @@
...
 			else //ingest, patch or inject
 				M.ForceContractDisease(D)
 
+	if(method == INJECT && iscarbon(M))
+		var/mob/living/carbon/C = M
+		if(C.get_blood_id() == "blood")
+			if(!data || !(data["blood_type"] in get_safe_blood(C.dna.blood_type)))
+				C.reagents.add_reagent("toxin", reac_volume * 0.5)
+			else
+				C.blood_volume = min(C.blood_volume + round(reac_volume, 0.1), BLOOD_VOLUME_MAXIMUM)
```
Comment (phil235) [inline code/modules/reagents/chemistry/reagents/other_reagents.dm]: injecting blood into a monkey/human that use actual blood reagent as blood replenish its volume.

File: code/modules/reagents/reagent_containers/blood_pack.dm [hunk 18]
```diff
@@ -4,19 +4,35 @@
...
 		update_icon()
 
 /obj/item/weapon/reagent_containers/blood/on_reagent_change()
+	if(reagents)
+		var/datum/reagent/blood/B = reagents.has_reagent("blood")
+		if(B && B.data && B.data["blood_type"])
+			blood_type = B.data["blood_type"]
+		else
+			blood_type = null
+	update_pack_name()
```
Comment (phil235) [inline code/modules/reagents/reagent_containers/blood_pack.dm]: the blood pack's name is automatically updated with the correct blood type in its name, unless someone manually labeled the pack (so traitors can still do traitor things).

File: code/modules/reagents/reagent_containers/blood_pack.dm [hunk 19]
```diff
@@ -49,9 +65,11 @@
 /obj/item/weapon/reagent_containers/blood/OMinus
 	blood_type = "O-"
 
+/obj/item/weapon/reagent_containers/blood/lizard
+	blood_type = "L"
```
Comment (phil235) [inline code/modules/reagents/reagent_containers/blood_pack.dm]: lizards now have a special blood type that only they use. No blood of that type is stored in medbay storage, dunno if we should or not.

File: code/modules/mob/living/carbon/human/blood.dm [hunk 20]
```diff
@@ -1,318 +0,0 @@
...
-****************************************************/
-//Blood levels
-var/const/BLOOD_VOLUME_NORMAL = 560
-var/const/BLOOD_VOLUME_SAFE = 501
-var/const/BLOOD_VOLUME_OKAY = 336
-var/const/BLOOD_VOLUME_BAD = 224
-var/const/BLOOD_VOLUME_SURVIVE = 122
-
-/mob/living/carbon/human/var/datum/reagents/vessel	//Container for blood and BLOOD ONLY. Do not transfer other chems here.
-/mob/living/carbon/human/var/pale = 0			//Should affect how mob sprite is drawn, but currently doesn't.
```
Comment (phil235) [inline code/modules/mob/living/carbon/human/blood.dm]: and fuck those vars too.

File: code/modules/mob/living/carbon/human/blood.dm [hunk 21]
```diff
@@ -1,318 +0,0 @@
...
-			if(brutedamage > 30)
-				blood_max += 0.5
-			if(brutedamage > 50)
-				blood_max += 1
-			if(brutedamage > 70)
-				blood_max += 2
-		if(bleedsuppress)
-			blood_max = 0
-		if(reagents.has_reagent("heparin") && getBruteLoss()) //Heparin is a powerful toxin that causes bleeding
-			blood_max += 3
```
Comment (phil235) [inline code/modules/mob/living/carbon/human/blood.dm]: no, we don't check for specific reagents every time we call Life(), this is stupid. I've moved the code into the on_mob_life() of heparin, nutriment, vitamin, and iron.

File: code/modules/reagents/chemistry/reagents/toxin_reagents.dm [hunk 22]
```diff
@@ -649,7 +649,7 @@
 /datum/reagent/toxin/heparin/on_mob_life(mob/living/M)
 	if(ishuman(M))
 		var/mob/living/carbon/human/H = M
-		H.blood_max += 2
+		H.bleed_rate = min(H.bleed_rate + 2, 8)
```
Comment (phil235) [inline code/modules/reagents/chemistry/reagents/toxin_reagents.dm]: this is technically a buff. the old code was nonsensical so there's no way for me to tell what was bug and what was intended. the bleeding var was increased here, but handle_blood() was reseting it every time, meaning the heparin code here did nothing. There was a special snowflake code for heparin in handle_blood() but the effect didn't match what's here. So there you go I guess.

File: code/modules/mob/living/carbon/human/species.dm [hunk 23]
```diff
@@ -731,6 +729,10 @@
 	return
 
 /datum/species/proc/handle_chemicals(datum/reagent/chem, mob/living/carbon/human/H)
+	if(chem.id == exotic_blood)
+		H.blood_volume = min(H.blood_volume + round(chem.volume, 0.1), BLOOD_VOLUME_MAXIMUM)
+		H.reagents.remove_reagent(chem.id)
+		return 1
```
Comment (phil235) [inline code/modules/mob/living/carbon/human/species.dm]: if a chemical in our reagents match our blood substance, we absorb it to add it to our blood volume.

**Comment by Incoming5643** [comment]
Cool and sorely needed, but:

> can't draw blood from mob if it's low on blood, so we can't empty a mob of its blood entirely

Why not?

**Comment by phil235** [comment]
@Incoming5643 It seems pretty cheap and unrealistic, being able to kill people that way.

File: code/game/objects/items/devices/radio/intercom.dm [hunk 24]
```diff
@@ -67,5 +67,5 @@
 		else
 			icon_state = "intercom"
 
-/obj/item/device/radio/intercom/rejects_blood()
-	return 1
+/obj/item/device/radio/intercom/add_mob_blood(mob/living/carbon/M)
+	return 0
```
Comment (AnturK) [inline code/game/objects/items/devices/radio/intercom.dm]: Why is this needed anyway ?

Comment (phil235) [inline code/game/objects/items/devices/radio/intercom.dm]: *shrug*

File: code/modules/mob/living/carbon/human/species_types.dm [hunk 25]
```diff
@@ -75,6 +75,13 @@
 
 	return randname
 
+/datum/species/lizard/on_species_gain(mob/living/carbon/C)
+	if(C.dna.blood_type != "L")
+		C.dna.blood_type = "L"
```
Comment (AnturK) [inline code/modules/mob/living/carbon/human/species_types.dm]: This could possibly be a generic species var ? Not that i expect we'll be adding many species with unique bloodtypes any time soon.

File: code/modules/mob/living/simple_animal/simple_animal.dm (hunk 14 above)
Comment (AnturK) [inline code/modules/mob/living/simple_animal/simple_animal.dm]: This will probably have some issues with player controlled simple animals. Hopefully nothing major.

**Comment by optimumtact** [comment]
Ready?

**Comment by phil235** [comment]
not yet.

**Comment by AnturK** [comment]
:+1: Could use a changelog on the drag bleed though.

**Comment by phil235** [comment]
should be ready now, unless people have more feedback.

**Comment by KorPhaeron** [comment]
How many tiles can you drag someone before it kills them?

**Comment by phil235** [comment]
@KorPhaeron Actually it's pretty much impossible to kill them by blood loss from dragging. Look at the code:
```
			var/brute_ratio = round(getBruteLoss()/maxHealth, 0.1)
			if(blood_volume && blood_volume > max(BLOOD_VOLUME_NORMAL*(1 - brute_ratio * 0.25), 0))//don't leave trail if blood volume below a threshold
				blood_volume = max(blood_volume - max(1, brute_ratio * 2), 0) 		
```
I erred wayyy into the safe side. A human with 100 brute damage and all its blood volume at the start will only be able get as low as 87.5% blood volume, and it will lose 1cL of blood for every tiles he's dragged on. So you can drag them over roughly 70 tiles before he stops shedding blood on the ground from the dragging. They will only be able to lose more blood from dragging if the get more brute damage.

**Comment by KorPhaeron** [comment]
Sounds good then
